### PR TITLE
GitHub Issue NOAA-EMC/GSI#335. Add several new functions from RAP/HRRR/RRFS application.

### LINKED ISSUE #335: Use 2m T/Q as background for surface observation for RRFS and remove the negative moisture in low level for RAP/HRRR
Add several new functions from RAP/HRRR/RRFS application:
    
    1) Add function to read and write 2m T and Q from fv3lam background
       and used 2m T and Q as background to calculate innovation for
       2m T/Q surface observations.
    
    2) Change the variable name for 2m T from th2 to t2m to indicate
       It is a sensible temperature now. It was a potential temperature.
       The converting between potential and sensible is in
       background IO (from ARW)
    
    3) Add function to remove the negative moisture analysis
       if background is not too dry (RH > 4%) for
       the lowest three model levels. This function is in the code to
       write analysis results for ARW and only used in RAP/HRRR.

### PR DESCRIPTION
**Pull request by hu5970**
This PR address issue #335:

1) Add function to read and write 2m T and Q from fv3lam background
   and used 2m T and Q as background to calculate innovation for
   2m T/Q surface observations.

2) Change the variable name for 2m T from th2 to t2m to indicate
   It is a sensible temperature now. It was a potential temperature.
   The converting between potential and sensible is in
   background IO (from ARW)

3) Add function to remove the negative moisture analysis
   if background is not too dry (RH > 4%) for
   the lowest three model levels. This function is in the code to
   write analysis results for ARW and only used in RAP/HRRR.

Note:
The anavinfo for RAP/HRRR need to be changed:
change:
  th2m      1      -1         2-m_T                th2m
to
  t2m       1      -1         2-m_T                t2m

Also, the results of the analysis for ARW background (RAP/HRRR) is changed.

### CODE REVIEW

File: src/gsi/cplr_wrwrfmassa.f90 [hunk 1]
```diff
@@ -764,12 +764,23 @@ subroutine wrwrfmassa_binary_wrf(this,mype)
...
+          ip1=i+1
+          do j=1,lat1
+  !  Convert 2m sensible T to potential T
+             jp1=j+1
+             all_loc(j,i,i_th2)=ges_t2m(jp1,ip1)*(r100/ges_ps(jp1,ip1))**rd_over_cp_mass
+          end do
+       end do
+    endif
     if (i_use_2mq4b > 0) then
+       call GSI_BundleGetPointer ( GSI_MetGuess_Bundle(it), 'q2m' ,  ges_q2,  istatus );ier=ier+istatus
```
Comment (guoqing-noaa) [inline src/gsi/cplr_wrwrfmassa.f90:783]: Just a question, so we don't get q2m before this PR?

File: src/gsi/gsd_update_mod.f90 [hunk 2]
```diff
@@ -530,14 +533,14 @@ subroutine gsd_update_th2(tinc,it)
...
-  real(r_kind),dimension(:,:  ),pointer:: ges_th2=>NULL()
+  real(r_kind),dimension(:,:  ),pointer:: ges_t2m=>NULL()
   real(r_kind),dimension(:,:,:),pointer:: ges_q  =>NULL()
 
 !*******************************************************************************
 !
 ! 2-m temperature
 !  do it=1,nfldsig
-     call gsi_bundlegetpointer(gsi_metguess_bundle(it),'th2m',ges_th2,ier)
+     call gsi_bundlegetpointer(gsi_metguess_bundle(it),'t2m',ges_t2m,ier)
```
Comment (guoqing-noaa) [inline src/gsi/gsd_update_mod.f90:543]: This change will also work for global applications if gsd_update_t2m() is called , right?

File: src/gsi/gsd_update_mod.f90 [hunk 3]
```diff
@@ -552,17 +555,15 @@ subroutine gsd_update_th2(tinc,it)
               if(ihaveq/=0) cycle
               dth2=tinc(i,j)/(one+fv*ges_q(i,j,1))
            endif
-!          Convert sensible temperature to potential temperature
-           work_prsl  = one_tenth*(aeta1_ll(1)*(r10*ges_ps(i,j)-pt_ll)+ &
-                                   aeta2_ll(1) + pt_ll)
-           work_prslk = (work_prsl/r100)**rd_over_cp_mass
-           ges_th2(i,j) = ges_th2(i,j) + dth2/work_prslk
+
+!        do not need to convert sensible temperature to potential temperature
```
Comment (guoqing-noaa) [inline src/gsi/gsd_update_mod.f90:559]: Line 559 may be removed

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 4]
```diff
@@ -1071,7 +1091,13 @@ subroutine read_fv3_netcdf_guess(fv3filenamegin)
        
       endif
 
-      call gsi_fv3ncdf2d_read(fv3filenamegin,it,ges_z)
+      call gsi_fv3ncdf2d_read(fv3filenamegin,it,ges_z,ges_t2m,ges_q2m)
+
+      if(i_use_2mq4b > 0 .and. i_use_2mt4b > 0 ) then
+! do not need to convert t2m to potentional T in ges_t2m
```
Comment (guoqing-noaa) [inline src/gsi/gsi_rfv3io_mod.f90:1097]: Line 1097 can be removed

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 5]
```diff
@@ -2147,8 +2192,13 @@ subroutine wrfv3_netcdf(fv3filenamegin)
 !   write out
     if (ier/=0) call die('get ges','cannot get pointers for fv3 met-fields, ier =',ier)
 
-    add_saved=.true.
+    if(i_use_2mq4b > 0 .and. i_use_2mt4b > 0 ) then
+! do not need to convert t2m from potentional T  to T in ges_t2m
```
Comment (guoqing-noaa) [inline src/gsi/gsi_rfv3io_mod.f90:2196]: Line 2196 can be removed

**Review by guoqing-noaa — APPROVED** [review]

**Review by TingLei-NOAA — APPROVED** [review]

File: src/gsi/gsd_update_mod.f90 (hunk 3 above)
Comment (hu5970) [reply src/gsi/gsd_update_mod.f90]: removed

File: src/gsi/gsi_rfv3io_mod.f90 (hunk 4 above)
Comment (hu5970) [reply src/gsi/gsi_rfv3io_mod.f90]: removed

**Comment by hu5970** [comment]
@MichaelLueken-NOAA Hi Mike, any comments on this PR?

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 6]
```diff
@@ -884,6 +892,10 @@ subroutine read_fv3_netcdf_guess(fv3filenamegin)
             ntracerio2d=ntracerio2d+1
           else if(trim(vartem)=='z') then
              write(6,*)'the metvarname ',trim(vartem),' will be dealt separately'
+          else if(trim(vartem)=='t2m') then
+             !write(6,*)'the metvarname ',trim(vartem),' will be dealt separately'
+          else if(trim(vartem)=='q2m') then
+             !write(6,*)'the metvarname ',trim(vartem),' will be dealt separately'
```
Comment (MichaelLueken-NOAA) [inline src/gsi/gsi_rfv3io_mod.f90:898]: Should the writes be committed out?  If they won't be used, please remove commented out lines.

**Review by MichaelLueken — COMMENTED** [review]
Hi Ming,

I have completed my visual inspection of the changes in this PR.  I see two commit messages.  To bring this back to one, while maintaining the latest authoritative repository changes please do the following:

1) git reset --hard HEAD~1
2) git remote add upstream https://github.com/NOAA-EMC/GSI.git
3) git remote update
4) git rebase upstream/master
5) Correct the conflict in src/gsi/setupt.f90 (conflict in REVISION HISTORY only)
6) git add src/gsi/setupt.f90
7) git rebase --continue
8) git push origin gsl_2mtq4bk --force

 Additionally, please add (i_kind) to the new integer declaration, and if commented out lines won't be used, please remove them.

File: src/gsi/gsi_rfv3io_mod.f90 (hunk 6 above)
Comment (hu5970) [reply src/gsi/gsi_rfv3io_mod.f90:898]: the lines are removed

File: src/gsi/cplr_wrwrfmassa.f90 [hunk 7]
```diff
@@ -1882,6 +1897,7 @@ subroutine wrwrfmassa_netcdf_wrf(this,mype)
   ! Declare local parameters
     character(len=*),parameter::myname='wrwrfmassa_netcdf'
     real(r_kind),parameter:: r225=225.0_r_kind
+    integer, parameter :: nlvl_qcheck=3
```
[the remark this answers is not in the document]
Comment (hu5970) [reply src/gsi/cplr_wrwrfmassa.f90:1900]: changed to:
integer(i_kind), parameter :: nlvl_qcheck=3_i_kind

**Comment by hu5970** [comment]
@MichaelLueken-NOAA Thanks for comments and instructions on syncing with the master. I have addressed all issues. Ming

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 8]
```diff
@@ -708,7 +714,7 @@ subroutine read_fv3_netcdf_guess(fv3filenamegin)
     use mpimod, only: mpi_comm_world
     use guess_grids, only:ges_prsi
     use gridmod, only: lat2,lon2,nsig,ijn,eta1_ll,eta2_ll,ijn_s
-    use constants, only: one,fv
+    use constants, only: one,fv,rd_over_cp_mass,r10,r1000
```
Comment (MichaelLueken) [inline src/gsi/gsi_rfv3io_mod.f90]: The variable, rd_over_cp_mass, was identified as unused while compiling in debug.  Please remove unused variable.

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 9]
```diff
@@ -2043,6 +2076,8 @@ subroutine wrfv3_netcdf(fv3filenamegin)
     use directDA_radaruse_mod, only: l_use_dbz_directDA
     use directDA_radaruse_mod, only: l_cvpnr, cvpnr_pval
     use gridmod,  only: eta1_ll,eta2_ll
+    use constants, only: one,rd_over_cp_mass,r10,r1000
```
Comment (MichaelLueken) [inline src/gsi/gsi_rfv3io_mod.f90]: The variable, rd_over_cp_mass, has been identified as being unused.  Please remove unused variable.

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 10]
```diff
@@ -2043,6 +2076,8 @@ subroutine wrfv3_netcdf(fv3filenamegin)
...
+    use mpimod, only: mype
```
Comment (MichaelLueken) [inline src/gsi/gsi_rfv3io_mod.f90]: The variable, mype, has been identified as being unused.  Please remove unused variable.

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 11]
```diff
@@ -2591,6 +2643,136 @@ subroutine gsi_fv3ncdf_writeuv_v1(grd_uv,ges_u,ges_v,add_saved,fv3filenamegin)
...
+!
+!   output argument list:
+!
+! attributes:
+!   language: f90
+!   machine:
+!
+!$$$ end documentation block
+
+    use mpimod, only: ierror,mpi_comm_world,npe,mpi_rtype,mype
```
Comment (MichaelLueken) [inline src/gsi/gsi_rfv3io_mod.f90]: The variable, npe, has been identified as unused.  Please remove unused variable.

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 12]
```diff
@@ -2591,6 +2643,136 @@ subroutine gsi_fv3ncdf_writeuv_v1(grd_uv,ges_u,ges_v,add_saved,fv3filenamegin)
...
+    use gridmod, only: strip
+    implicit none
+
+    real(r_kind)   ,intent(in   ) :: var(lat2,lon2)
+    logical        ,intent(in   ) :: add_saved
+    character(*)   ,intent(in   ) :: varname
+    type (type_fv3regfilenameg),intent (in) :: fv3filenamegin
+
+    integer(i_kind) :: VarId,gfile_loc
+    integer(i_kind) i,j,mm1
```
Comment (MichaelLueken) [inline src/gsi/gsi_rfv3io_mod.f90]: The variable, j, has been identified as unused.  Please remove unused variable.

File: src/gsi/gsd_update_mod.f90 [hunk 13]
```diff
@@ -516,7 +518,8 @@ subroutine gsd_update_th2(tinc,it)
   use jfunc, only:  tsensible
   use constants, only: zero,one,fv,rd_over_cp_mass,one_tenth
```
Comment (MichaelLueken) [inline src/gsi/gsd_update_mod.f90]: The variable, rd_over_cp_mass, has been identified as unused.  Please remove unused variable.

File: src/gsi/gsd_update_mod.f90 [hunk 14]
```diff
@@ -516,7 +518,8 @@ subroutine gsd_update_th2(tinc,it)
...
   use gridmod, only: lat2,lon2,aeta1_ll,pt_ll,aeta2_ll
```
Comment (MichaelLueken) [inline src/gsi/gsd_update_mod.f90]: The variables, aeta1_ll, pt_ll, and aeta2_ll, have been identified as unused. Please remove unused variables.

File: src/gsi/gsd_update_mod.f90 [hunk 15]
```diff
@@ -516,7 +518,8 @@ subroutine gsd_update_th2(tinc,it)
...
-! use guess_grids, only: nfldsig
+  use guess_grids, only:ges_prsi
```
Comment (MichaelLueken) [inline src/gsi/gsd_update_mod.f90]: The variable, ges_prsi, has been identified as unused.  Please remove unused variable.

File: src/gsi/gsd_update_mod.f90 [hunk 16]
```diff
@@ -530,14 +533,14 @@ subroutine gsd_update_th2(tinc,it)
   real(r_kind) :: dth2, work_prsl,work_prslk
```
Comment (MichaelLueken) [inline src/gsi/gsd_update_mod.f90]: The variables, work_prsl and work_prslk, have been identified as unused.  Please remove unused variables.

**Review by MichaelLueken — COMMENTED** [review]
@hu5970, I have completed my review of your changes.  The visual inspection of the changes was completed and everything looks good.  I ran the regression tests (both standard and debug) and they all passed without issue.  I compiled the code on Hera using GNU compilers and ran the regression tests there.  All tests successfully passed.  The only thing that was noted were unused variables while compiling in debug mode.  I have noted the unused variables that the compiler flagged in my review.  Once they have been removed, I should be able to send this work out to the review committee once the current work has been merged to the authoritative develop branch (on Friday).

File: src/gsi/gsi_rfv3io_mod.f90 (hunk 8 above)
Comment (hu5970) [reply src/gsi/gsi_rfv3io_mod.f90]: removed

File: src/gsi/gsi_rfv3io_mod.f90 (hunk 9 above)
Comment (hu5970) [reply src/gsi/gsi_rfv3io_mod.f90]: removed "rd_over_cp_mass,r10,r1000"

File: src/gsi/gsi_rfv3io_mod.f90 (hunk 10 above)
Comment (hu5970) [reply src/gsi/gsi_rfv3io_mod.f90]: removed

File: src/gsi/gsi_rfv3io_mod.f90 (hunk 11 above)
Comment (hu5970) [reply src/gsi/gsi_rfv3io_mod.f90]: removed

File: src/gsi/gsi_rfv3io_mod.f90 (hunk 12 above)
Comment (hu5970) [reply src/gsi/gsi_rfv3io_mod.f90]: removed

File: src/gsi/gsd_update_mod.f90 (hunk 13 above)
Comment (hu5970) [reply src/gsi/gsd_update_mod.f90]: removed

File: src/gsi/gsd_update_mod.f90 (hunk 14 above)
Comment (hu5970) [reply src/gsi/gsd_update_mod.f90]: removed

File: src/gsi/gsd_update_mod.f90 (hunk 15 above)
Comment (hu5970) [reply src/gsi/gsd_update_mod.f90]: removed.

File: src/gsi/gsd_update_mod.f90 (hunk 16 above)
Comment (hu5970) [reply src/gsi/gsd_update_mod.f90]: removed

**Comment by hu5970** [comment]
@MichaelLueken-NOAA Thanks for your comments. Please let me know if I can help with something. Ming

**Comment by MichaelLueken** [comment]
@hu5970 Following the merge of Li's FGAT for regional DA update on Friday, there is now a conflict with gsi_rfv3io_mod.f90.  To correct this:

```
1) git clone --recursive git@github.com:hu5970/GSI.git update
2) cd update
3) git checkout gsl_2mtq4bk
4) git submodule update
5) git remote add upstream https://github.com/NOAA-EMC/GSI.git
6) git remote update
7) git rebase upstream/develop
8) Address issues with src/gsi/gsi_rfv3io_mod.f90
9) git add src/gsi/gsi_rfv3io_mod.f90
10) git submodule update
11) git commit --amend
12) git push origin gsl_2mtq4bk --force
```

If you have any questions, please let me know.

**Comment by hu5970** [comment]
@MichaelLueken-NOAA I have solved the conflict in gsi_rfv3io_mod.f90. Thanks,

**Comment by MichaelLueken** [comment]
The due date for the review committee has past with no feedback, so I will now give final approval to these changes and merge them to the authoritative develop branch.

**Pull request merged**